Add release note for kubestash v2024.6.4 (#435)
appscode/blog: content/post/kubestash-v2024.6.4/index.md

---
title: Introducing KubeStash v2024.6.4
date: "2024-06-05"
weight: 10
authors:
- Md Ishtiaq Islam
tags:
- backup
- disaster-recovery
- kubernetes
- kubestash
- manifest-backup
- manifest-restore
- restore
---

We are pleased to announce the release of [KubeStash v2024.6.4](https://kubestash.com/docs/v2024.6.4/setup/), packed with new features and important bug fixes. You can check out the full changelog [HERE](https://github.com/kubestash/CHANGELOG/blob/master/releases/v2024.6.4/README.md). In this post, we'll highlight the key updates.

### New Features

Here, we are going to highlight the new features that have been introduced in this release.

#### Workload Manifest Backup & Restore

> We've introduced functionality to backup and restore workload manifests within the `workload-addon`. These tasks encompass backing up and restoring the manifests of the workload, including their associated volumes, service account (if utilized), and service (only the headless service employed in StatefulSet).

Reviewer: ```suggestion
We've introduced functionality to backup and restore workload (Deployment/StatefulSet/DaemonSet) manifests within the `workload-addon`. The new `manifest-backup` and `manifest-restore` tasks of the `workload-addon` enable backing up and restoring of workload manifests, including their associated volumes, service account (if used), and service (in case of StatefulSet).
```

      namespace: demo
      name: s3-storage
    retentionPolicy:
      name: demo-retention
      namespace: demo
  sessions:
  - name: workload-backup
    sessionHistoryLimit: 3
    scheduler: 
      schedule: "*/5 * * * *"
      jobTemplate:
        backoffLimit: 1
    repositories:
    - name: demo-storage
      backend: s3-storage
      directory: /sts
      encryptionSecret:
        name: encry-secret
        namespace: demo
    addon:
      name: workload-addon
      tasks:
      - name: manifest-backup
    retryConfig:
      maxRetry: 2
      delay: 1m
```

Here is the example of `RestoreSession` that will restore the StatefulSet manifest:

```yaml
apiVersion: core.kubestash.com/v1alpha1
kind: RestoreSession
metadata:
  name: sample-restore
  namespace: demo
spec:
  manifestOptions:
    restoreNamespace: dev
  dataSource:
    repository: demo-storage
    snapshot: latest
    encryptionSecret:
      name: encry-secret
      namespace: demo
  addon:
    name: workload-addon
    tasks:
    - name: manifest-restore
```

Here, we can configure in which namespace we want to restore our workload by providing the namespace name in `spec.manifestOptions.restoreNamespace`.

> 

Reviewer: Use a proper title.

> We can also use these manifest tasks in case of application level backup and restore. To do so, we need to configure `BackupConfiguration` in the following manner:

Reviewer: ```suggestion
Here is an example BackupConfiguration for backing up both the manifests and data of a StatfeulSet in the same session: 

```

Here is an example of a `BackupConfiguration` for backing up both the manifests and data of a `StatefulSet` in the same session:

```yaml
apiVersion: core.kubestash.com/v1alpha1
kind: BackupConfiguration
metadata:
  name: sample-backup
  namespace: demo
spec:
  target:
    apiGroup: apps
    kind: StatefulSet
    name: sample-sts
    namespace: demo
  backends:
  - name: s3-storage
    storageRef:
      namespace: demo
      name: s3-storage
    retentionPolicy:
      name: demo-retention
      namespace: demo
  sessions:
  - name: workload-backup
    sessionHistoryLimit: 3
    scheduler:
      schedule: "*/5 * * * *"
      jobTemplate:
        backoffLimit: 1
    repositories:
    - name: demo-storage
      backend: s3-storage
      directory: /sts
      encryptionSecret:
        name: encry-secret
        namespace: demo
    addon:
      name: workload-addon
      tasks:
      - name: manifest-backup
      - name: logical-backup
        params:
          paths: /source/data
          exclude: /source/data/tmp
    retryConfig:
      maxRetry: 2
      delay: 1m
```

> Here's an example of a `RestoreSession` that restores `StatefulSet` manifests and then restores its data:

Reviewer: Here is an example of a RestoreSession that first redeploys the StatefulSet from the backed-up manifests and then restores the data into it:

metadata:
  name: sample-restore
  namespace: demo
spec:
  manifestOptions:
    restoreNamespace: dev
  dataSource:
    repository: demo-storage
    snapshot: latest
    encryptionSecret:
      name: encry-secret
      namespace: demo
  addon:
    name: workload-addon
    tasks:
    - name: manifest-restore
    - name: logical-backup-restore
```

In this case, a helper `RestoreSession` will be created in the same namespace of the applied `RestoreSession` to restore the data of the workload.

> #### MySQL, MariaDB, MongoDB & PostgreSQL Application Level Backup & Restore
> 
> We've added support application level backup and restore for `MySQL`, `MariaDB`, `MongoDB` and `PostgreSQL`. We'll be backing up both the database manifests and their associated data. When it comes to restoration, the manifests will be restored first, followed by the data.
> 
> Here is an example of `BackupConfiguration` for MySQL:

Reviewer: ```suggestion
Here is an example of `BackupConfiguration` for backing up both the manifests and data (dump) of a MySQL database in the same session:
```

```yaml
apiVersion: core.kubestash.com/v1alpha1
kind: BackupConfiguration
metadata:
  name: mysql-backup
  namespace: demo
spec:
  target:
    apiGroup: kubedb.com
    kind: MySQL
    namespace: demo
    name: mysql-backup
  sessions:
  - name: frequent-backup
    sessionHistoryLimit: 3
    scheduler:
      schedule: "*/5 * * * *"
      successfulJobsHistoryLimit: 1
      failedJobsHistoryLimit: 1
      jobTemplate:
        backoffLimit: 1
    repositories:
    - name: mysql-storage
      directory: /mysql
      encryptionSecret:
        name: encry-secret
        namespace: demo
    addon:
      name: mysql-addon
      tasks:
      - name: logical-backup
      - name: manifest-backup
    retryConfig:
      maxRetry: 2
      delay: 1m
```

> Here's an example of a `RestoreSession` that restores `MySQL` manifests and then restores its data:

Reviewer: Here is an example of a RestoreSession that first redeploys the MySQL database from the backed-up manifests and then restores the data into it:
